augur/dates.py

```diff
@@ -1,73 +1,271 @@
+
+
+def iso_to_numeric(date_in:str, ambiguity_resolver:str=None):
+    """Convert ISO 8601 date string to numeric, resolving any ambiguity detected by explicit 'X' characters or missing date parts.
+
+    Parameters
+    ----------
+    date_in
+        Date string in ISO 8601 format.
+
+    ambiguity_resolver
+        None: Assume the given ISO 8601 date string is an exact date.
+        'min': Resolve to minimum of ambiguous range.
+        'max': Resolve to maximum of ambiguous range.
+
+    >>> round(iso_to_numeric('2018-01-01'), 3)
+    2018.001
+    >>> round(iso_to_numeric('2018-03-25'), 2)
+    2018.23
+    >>> round(iso_to_numeric('2018-03', ambiguity_resolver='min'), 3)
+    2018.163
+    >>> round(iso_to_numeric('2018-03', ambiguity_resolver='max'), 3)
+    2018.245
+    >>> round(iso_to_numeric('2018', ambiguity_resolver='min'), 3)
+    2018.001
+    >>> round(iso_to_numeric('2018', ambiguity_resolver='max'), 3)
+    2018.999
+    >>> iso_to_numeric('2018-03-XX')
+    Traceback (most recent call last):
+      ...
+    ValueError: invalid literal for int() with base 10: 'XX'
+    >>> round(iso_to_numeric('2018-03-XX', ambiguity_resolver='min'), 3)
+    2018.163
+    >>> round(iso_to_numeric('2018-03-XX', ambiguity_resolver='max'), 3)
+    2018.245
+    >>> round(iso_to_numeric('2018-XX-25', ambiguity_resolver='min'), 3)
+    2018.067
+    >>> round(iso_to_numeric('2018-XX-25', ambiguity_resolver='max'), 3)
+    2018.982
+    >>> round(iso_to_numeric('201X-XX-XX', ambiguity_resolver='min'), 3)
+    2010.001
+    >>> round(iso_to_numeric('201X-XX-XX', ambiguity_resolver='max'), 3)
+    2019.999
     """
-    # date is numeric
+    date_parts = date_in.split('-', maxsplit=2)
+    # TODO: resolve partial month/day ambiguity eg. 2018-1X-XX, 2018-10-3X
+    if ambiguity_resolver is None:
+        year = int(date_parts[0])
+        month = int(date_parts[1])
+        day = int(date_parts[2])
+    elif ambiguity_resolver == 'min':
+        year = int(date_parts[0].replace('X', '0'))
+        month = int(date_parts[1]) if len(date_parts) > 1 and date_parts[1].isnumeric() else 1
+        day = int(date_parts[2]) if len(date_parts) > 2 and date_parts[2].isnumeric() else 1
+    elif ambiguity_resolver == 'max':
+        year = int(date_parts[0].replace('X', '9'))
+        month = int(date_parts[1]) if len(date_parts) > 1 and date_parts[1].isnumeric() else 12
+        if len(date_parts) == 3 and date_parts[2].isnumeric():
+            day = int(date_parts[2])
+        else:
+            if month in {1,3,5,7,8,10,12}:
+                day = 31
+            elif month == 2:
+                day = 28
+            else:
+                day = 30
     try:
-        return float(date)
+        return date_to_numeric_capped(datetime.date(year, month, day))
     except ValueError:
-        pass
+        # catches month/day out of bounds errors
+        raise InvalidDateFormat
```

Comment:
Should this exception object get a similar message to the other places we throw the same type?

Reply:
Yes, done: 468f3b46